### PR TITLE
Issue #2 added imagecarousel template

### PR DESCRIPTION
**Pull request by Ben67366**
#2 Tested, and it works.

### CODE REVIEW

**Comment by dlemstra** [comment]
Thanks for helping me out with this. Will do a proper review tomorrow.

File: src/LineBot/Messages/Template/IImageCarouselColumn.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+// License for the specific language governing permissions and limitations
+// under the License.
+
+using System;
+using System.Collections.Generic;
+
+namespace Line
+{
+    /// <summary>
+    /// Encapsulates a carousel column.
```
Comment (dlemstra) [inline src/LineBot/Messages/Template/IImageCarouselColumn.cs]: `image carousel`

File: src/LineBot/Messages/Template/IImageCarouselColumn.cs [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    /// </summary>
+    public interface IImageCarouselColumn
+    {
+        /// <summary>
+        /// Gets the image url for the thumbnail.
+        /// <para>Protocol: HTTPS</para>
+        /// <para>Format: JPEG or PNG</para>
+        /// <para>Max url length: 1000 characters</para>
+        /// <para>Aspect ratio: 1:1.51</para>
```
Comment (dlemstra) [inline src/LineBot/Messages/Template/IImageCarouselColumn.cs]: This should be: `Aspect ratio: 1:1` according to the documentation.

File: src/LineBot/Messages/Template/IImageCarouselColumn.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        /// <para>Max width: 1024px</para>
+        /// <para>Max size: 1 MB</para>
+        /// </summary>
+        Uri ImageUrl { get; }
+
+        /// <summary>
+        /// Gets Action when image is tapped
```
Comment (dlemstra) [inline src/LineBot/Messages/Template/IImageCarouselColumn.cs]: Can you change this to: `Gets the action when image is tapped.`

File: src/LineBot/Messages/Template/ImageCarouselColumn.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Newtonsoft.Json;
+
+namespace Line
+{
+    /// <summary>
+    /// Encapsulates a carousel column.
```
Comment (dlemstra) [inline src/LineBot/Messages/Template/ImageCarouselColumn.cs]: `image carousel`

File: src/LineBot/Messages/Template/ImageCarouselColumn.cs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+            {
+                return _imageUrl;
+            }
+
+            set
+            {
+                if (value != null)
+                {
+                    if (!"https".Equals(value.Scheme, StringComparison.OrdinalIgnoreCase))
+                        throw new InvalidOperationException("The thumbnail url should use the https scheme.");
```
Comment (dlemstra) [inline src/LineBot/Messages/Template/ImageCarouselColumn.cs]: Can you change `The thumbnail url` to `The image url`.

File: src/LineBot/Messages/Template/ImageCarouselColumn.cs [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+                        throw new InvalidOperationException("The thumbnail url cannot be longer than 1000 characters.");
+                }
+
+                _imageUrl = value;
+            }
+        }
+
+        /// <summary>
+        /// Gets or sets the actions when tapped.
+        /// <para>Max: 4</para>
```
Comment (dlemstra) [inline src/LineBot/Messages/Template/ImageCarouselColumn.cs]: Can you change this to `Gets or sets the action when image is tapped.` and remove the `Max` para?

File: tests/LineBot.Tests/Messages/Template/ImageCarouselTemplateTests/TheConstructor.cs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using Newtonsoft.Json;
+
+namespace Line.Tests.Messages.Template
+{
+    public partial class ImageCarCarouselTemplateTests
```
Comment (dlemstra) [inline tests/LineBot.Tests/Messages/Template/ImageCarouselTemplateTests/TheConstructor.cs]: Looks like a copy past mistake :)

File: src/LineBot/Messages/Template/ImageCarouselTemplate.cs [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+namespace Line
+{
+    /// <summary>
+    /// Encapsulates a image carousel template.
+    /// </summary>
+    public sealed class ImageCarouselTemplate : IImageCarouselTemplate
+    {
+        private IEnumerable<IImageCarouselColumn> _columns;
+
+        [JsonProperty("type")]
```
Comment (dlemstra) [inline src/LineBot/Messages/Template/ImageCarouselTemplate.cs]: Can you put this inside a pragma block to fix the stylecop compile error? `#pragma warning disable 0414 // Suppress value is never used.` should be added the same was this is done inside `CarouselTemplate`.

File: tests/LineBot.Tests/Messages/Template/ImageCarouselTemplateTests/TheConstructor.cs (hunk 7 above)
Comment (Ben67366) [reply tests/LineBot.Tests/Messages/Template/ImageCarouselTemplateTests/TheConstructor.cs]: Oh. lol.

**Comment by dlemstra** [comment]
Thanks for making the fixes @Ben67366. Could you please sign the CLA?

**Pull request closed without merging**